src/transport/tcp/tcp_connection.cpp

```diff
@@ -90,13 +96,57 @@ namespace libp2p::transport {
   void TcpConnection::connect(
       const TcpConnection::ResolverResultsType &iterator,
       TcpConnection::ConnectCallbackFunc cb) {
-    boost::asio::async_connect(socket_, iterator,
-                               [self{shared_from_this()}, cb{std::move(cb)}](
-                                   auto &&ec, auto &&endpoint) {
-                                 self->initiator_ = true;
-                                 cb(std::forward<decltype(ec)>(ec),
-                                    std::forward<decltype(endpoint)>(endpoint));
-                               });
+    connect(iterator, std::move(cb), std::chrono::milliseconds::zero());
+  }
+
+  void TcpConnection::connect(
+      const TcpConnection::ResolverResultsType &iterator,
+      ConnectCallbackFunc cb, std::chrono::milliseconds timeout) {
+    if (timeout > std::chrono::milliseconds::zero()) {
+      connecting_with_timeout_ = true;
+      deadline_timer_.expires_from_now(
+          boost::posix_time::milliseconds(timeout.count()));
+      deadline_timer_.async_wait([self{shared_from_this()},
+                                  cb](const boost::system::error_code &error) {
+        bool expected = false;
+        if (self->connection_phase_done_.compare_exchange_strong(expected,
+                                                                 true)) {
+          if (not error) {
+            // timeout happened, timer expired before connection was
+            // established
+            cb(boost::system::error_code{boost::system::errc::timed_out,
```

Comment:
I would suggest to create custom error for such case with more explicit explanation of the reason why connection failed

Reply:
I agree - this could be better. The current approach is defined due to interfaces limitations of boost::async_connect callback, which is required to consume boost::error_code. The handcrafted error_code object which is used now is well-formed and going to be served the same way as the error happened as the result of async_connect.